pkg/certificate/configuration.go

```diff
@@ -2,169 +2,453 @@ package certificate
+		certificateChain.CertificatesIssued[serviceHostname].CACertPEM[caBundleName] = config.CABundle
+		key := newObjectKey("Secret", serviceNamespace, serviceName)
+		if _, found := objects[key]; !found {
+			objects[key] = &keyedObject{key, nil}
 		}
+	}
+}
 
-		err = m.client.Update(context.TODO(), webhook)
-		if err != nil {
-			return err
+// mapWebhookToChain maps a webhook object from certificate chain data.
+func mapWebhookFromChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	clientConfigList := clientConfigMap(object.kobject)
+	for name, config := range clientConfigList {
+		serviceHostname := serviceHostname(config.Service.Name, config.Service.Namespace)
+		certificateIssue := certificateChain.CertificatesIssued[serviceHostname]
+		if certificateIssue == nil {
+			continue
 		}
-		return nil
-	})
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to update webhook CABundle")
+		caBundleName := caBundleName(object.key.String(), name)
+		caBundle := certificateIssue.CACertPEM[caBundleName]
+		if caBundle == nil {
+			continue
+		}
+		config.CABundle = caBundle
+	}
+}
+
+// mapWebhookToChain maps a secret object to certificate chain data.
+func mapSecretToChain(object *keyedObject, objects objectMap, certificateChain *chain.CertificateChainData) {
+	if object.key.NamespacedName.String() == certificateChain.CA.Name {
+		mapCASecretToChain(object, certificateChain)
+		return
+	}
+	mapServiceSecretToChain(object, certificateChain)
+}
+
+// mapWebhookToChain maps a secret object from certificate chain data.
+func mapSecretFromChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	if object.key.NamespacedName.String() == certificateChain.CA.Name {
+		mapCASecretFromChain(object, certificateChain)
+		return
+	}
+	mapServiceSecretFromChain(object, certificateChain)
+}
+
+func mapServiceSecretToChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	secret := object.kobject.(*corev1.Secret)
+	key := secret.Data[corev1.TLSPrivateKeyKey]
+	cert := secret.Data[corev1.TLSCertKey]
+	if key == nil || cert == nil {
+		return
+	}
+
+	name := serviceHostname(object.key.NamespacedName.Name, object.key.NamespacedName.Namespace)
+
+	certificateChain.CertificatesIssued[name].KeyPEM = key
+	certificateChain.CertificatesIssued[name].CertPEM = cert
+}
+
+func mapCASecretToChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	secret := object.kobject.(*corev1.Secret)
+	key := secret.Data[CAPrivateKeyKey]
+	cert := secret.Data[CACertKey]
+	if key == nil || cert == nil {
+		return
+	}
+	certificateChain.CA.KeyPEM = key
+	certificateChain.CA.CertPEM = cert
+}
+
+func mapServiceSecretFromChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	secret := object.kobject.(*corev1.Secret)
+	name := serviceHostname(object.key.NamespacedName.Name, object.key.NamespacedName.Namespace)
+	bundle := certificateChain.CertificatesIssued[name]
```

Comment:
Can we rename "bundle" ? the bundle is the PEM that contains more than one certificate so that's why is named bundle, in this case this is a KeyPair or KeyCert, maybe we can name this serviceKeyCert since we are using a serviceHostName to retrieve it.